sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/SupportsRuntimeFiltering.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.read;
+
+import org.apache.spark.annotation.Experimental;
+import org.apache.spark.sql.connector.expressions.NamedReference;
+import org.apache.spark.sql.sources.Filter;
+
+/**
+ * A mix-in interface for {@link Scan}. Data sources can implement this interface if they can
+ * filter initially planned {@link InputPartition}s using predicates Spark infers at runtime.
+ * <p>
+ * Note that Spark will push runtime filters only if they are beneficial.
+ *
+ * @since 3.2.0
+ */
+@Experimental
+public interface SupportsRuntimeFiltering extends Scan {
+  /**
+   * Returns attributes this scan can be filtered by at runtime.
+   * <p>
+   * Spark will call {@link #filter(Filter[], boolean)} if it can derive a runtime
+   * predicate for any of the filter attributes.
+   */
+  NamedReference[] filterAttributes();
+
+  /**
+   * Filters this scan using runtime filters.
+   * <p>
+   * The provided expressions must be interpreted as a set of filters that are ANDed together.
+   * Implementations may use the filters to prune initially planned {@link InputPartition}s.
+   * <p>
+   * The scan must always preserve the original data distribution during runtime filtering.
+   * It is allowed to change the number of partitions iff the `canChangeNumPartitions` flag is true.
+   * During runtime filtering, the scan may detect that some {@link InputPartition}s have no
+   * matching data. It can omit such partitions entirely only if `canChangeNumPartitions` is true.
+   * If `canChangeNumPartitions` is false, the scan can replace the initially planned
+   * {@link InputPartition}s that have no matching data with empty {@link InputPartition}s.
+   * <p>
+   * Note that Spark will call {@link Scan#toBatch()} again after filtering the scan at runtime.
+   *
+   * @param filters data source filters used to filter the scan at runtime
+   * @param canChangeNumPartitions a flag whether the scan can change the number of partitions
+   */
+  void filter(Filter[] filters, boolean canChangeNumPartitions);
```

Comment:
It would be much cleaner to avoid this boolean flag. I'd appreciate if everyone can think this through with me.

The current proposal is to require data sources to keep the original distribution during runtime filtering. The question is whether changing the number of tasks would break anything.

I can think of these scenarios:
- Joins without shuffles with `SinglePartition` partitioning. Such cases are easy to handle. If runtime filtering filters out the only split, we can still emit an empty task and hence guarantee the exact same partitioning.
- Bucketed/storage-partitioned joins for v2 tables. We don't know the exact plan now (@sunchao is working on a proposal). There are ideas where this would not really matter. It seems like we can still address it there. 
- Some streaming queries maybe? Streaming joins? But we don't really support predicate pushdown in streaming so I am not sure whether we need to care about that.

cc @viirya @dongjoon-hyun @sunchao @cloud-fan @rdblue @HyukjinKwon, I'd love to hear your thoughts.

Reply:
I did take a look at the code one more time. I think `SupportsReportPartitioning` was mostly added to avoid shuffling data during aggregates where the number of tasks does not really matter. As long as we require data sources to keep the original distribution during runtime filtering, we should be good.

If we want to be absolutely safe, we can say that data sources that implement `SupportsReportPartitioning`, must preserve not only distribution but also partitioning (including the number of tasks). It is most likely that interface will change as we discussed earlier. We can reconsider the requirement once the redesign happens. In all other cases, it is safe to change the number of tasks during runtime filtering.

Reply:
I am going to update this PR and remove the flag to have a simple API.

Reply:
+1 to removing this. I agree with @aokolnychyi's analysis.

The main problem is with bucketed joins, but I think this would be handled by Spark. In the v1 bucketed join path, Spark already performs the work to group files into a task per bucket. It isn't unreasonable that the v2 work for bucketed/storage-partitioned joins will do the same thing rather than trying to push the clustered distribution down into sources to plan around. That would also put more unnecessary complexity on sources that should be shared across all sources by implementing it in Spark.

Reply:
So I'm curious as to what the overhead of having empty partitions is it seems like it wouldn't be too bad but from this writing it sounds like my intuition isn't in line with y'alls understanding.

Reply:
This looks OK to me too. The current design of storage-partitioned join checks both the number of partitions as well as the partition values, e.g., `[1, 2]` is considered to be co-partitioned with `[1, 2]` but not `[1, 3]`. I can't image a case where the number of tasks remains the same after runtime filtering but the partition values have changed though.